crypto/sm2/sm2_sign.c

```diff
@@ -0,0 +1,193 @@
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/sm2.h>
+#include <openssl/evp.h>
+
+static BIGNUM *compute_msg_hash(const EVP_MD *digest,
+                                const EC_KEY *key,
+                                const char *user_id,
+                                const uint8_t *msg, size_t msg_len)
+{
+    EVP_MD_CTX *hash = EVP_MD_CTX_new();
+    const int md_size = EVP_MD_size(digest);
+    uint8_t za[md_size];
+    BIGNUM *e = NULL;
+
+    if (SM2_compute_za(za, digest, user_id, key) == 0)
+        goto done;
+
+    if (EVP_DigestInit(hash, digest) == 0)
+        goto done;
+
+    if (EVP_DigestUpdate(hash, za, md_size) == 0)
+        goto done;
+
+    if (EVP_DigestUpdate(hash, msg, msg_len) == 0)
+        goto done;
+
+    // reuse za buffer to hold H(ZA || M)
+    if (EVP_DigestFinal(hash, za, NULL) == 0)
+        goto done;
+
+    e = BN_bin2bn(za, md_size, NULL);
+
+ done:
+    EVP_MD_CTX_free(hash);
+    return e;
+}
+
+ECDSA_SIG *SM2_do_sign(const EC_KEY *key,
+                       const EVP_MD *digest,
+                       const char *user_id, const uint8_t *msg, size_t msg_len)
+{
+    ECDSA_SIG *sig = NULL;
+    BIGNUM *k = BN_new();
+    BIGNUM *rk = BN_new();
+    BIGNUM *e = NULL;
+    BIGNUM *r = BN_new();
+    BIGNUM *s = BN_new();
+    BIGNUM *x1 = BN_new();
+    BIGNUM *tmp = BN_new();
+    BN_CTX *ctx = BN_CTX_new();
+
+    const BIGNUM *dA = EC_KEY_get0_private_key(key);
+    const EC_GROUP *group = EC_KEY_get0_group(key);
+    const BIGNUM *order = EC_GROUP_get0_order(group);
+    EC_POINT *kG = EC_POINT_new(group);
+
+    /*
+       A1: set M~=ZA || M
+       A2: calculate e=Hv(M~)
+       A3: pick a random number k in [1, n-1] via a random number generator
+       A4: calculate the elliptic curve point (x1, y1)=[k]G
+       A5: calculate r=(e+x1) modn, return to A3 if r=0 or r+k=n
+       A6: calculate s=((1+dA)^(-1)*(k-r*dA)) modn, return to A3 if s=0
+       A7: the digital signature of M is (r, s)
+     */
+
+    e = compute_msg_hash(digest, key, user_id, msg, msg_len);
+    if (e == NULL)
+        goto done;
+
+    for (;;) {
+        BN_priv_rand_range(k, order);
+
+        if (EC_POINT_mul(group, kG, k, NULL, NULL, ctx) == 0)
+            goto done;
+
+        if (EC_POINT_get_affine_coordinates_GFp(group, kG, x1, NULL, ctx) == 0)
+            goto done;
+
+        if (BN_mod_add(r, e, x1, order, ctx) == 0)
+            goto done;
+
+        // try again if r == 0 or r+k == n
+        if (BN_is_zero(r))
+            continue;
+
+        BN_add(rk, r, k);
+
+        if (BN_cmp(rk, order) == 0)
+            continue;
+
+        BN_add(s, dA, BN_value_one());
+        BN_mod_inverse(s, s, order, ctx);
+
+        BN_mod_mul(tmp, dA, r, order, ctx);
+        BN_sub(tmp, k, tmp);
+
+        BN_mod_mul(s, s, tmp, order, ctx);
+
+        sig = ECDSA_SIG_new();
```

Comment:
This allocation could fail and needs to be checked for NULL

Reply:
Fixed